File: app/src/main/java/emreaktrk/edgecontact/ui/edge/contact/ContactView.java
package emreaktrk.edgecontact.ui.edge.contact;


import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.support.design.widget.FloatingActionButton;
import android.util.AttributeSet;
import android.view.View;
import emreaktrk.edgecontact.R;

public final class ContactView extends FloatingActionButton implements View.OnClickListener {

    private OnClickListener mListener;
    private Contact mContact;

    public ContactView(Context context, AttributeSet attrs) {
        super(context, attrs);

        init();
    }

    public ContactView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        init();
    }

    private void init() {
        setOnClickListener(this);
    }

    public void setOnClickListener(OnClickListener listener) {
        mListener = listener;
    }

    @Override
    public void onClick(View view) {
        if (mListener == null) {
            return;
        }

        if (hasContact()) {
            mListener.onCallClicked(mContact, this);
        } else {
            mListener.onAddClicked(this);
        }
    }

    public void setContact(@Nullable Contact contact) {
        mContact = contact;
        update();
    }

    private void update() {
        if (hasContact()) {
            apply();
            return;
        }

        clear();
    }

    public boolean hasContact() {
        return mContact != null;
    }

    @UiThread
    private void clear() {
        post(() -> setImageResource(R.drawable.ic_add));
    }

    @UiThread
    private void apply() {
        final Drawable drawable = mContact.hasPhoto() ? mContact.roundedPhoto(getContext()) : mContact.letterDrawable();

        post(() -> setImageDrawable(drawable));
    }

    public interface OnClickListener {

        void onCallClicked(Contact contact, View view);

        void onAddClicked(View view);
    }
}

File: app/src/main/java/emreaktrk/edgecontact/ui/edge/contact/ContactView.kt
package emreaktrk.edgecontact.ui.edge.contact

import android.content.Context
import android.support.annotation.UiThread
import android.support.design.widget.FloatingActionButton
import android.util.AttributeSet
import android.view.View
import emreaktrk.edgecontact.R

class ContactView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : FloatingActionButton(context, attrs, defStyleAttr), View.OnClickListener {

    private var listener: OnClickListener? = null
    private var contact: Contact? = null

    init {
        setOnClickListener(this)
    }

    fun setOnClickListener(listener: OnClickListener?) {
        this.listener = listener
    }

    override fun onClick(view: View) {
        if (listener == null) {
            return
        }
        if (hasContact()) {
            listener!!.onCallClicked(contact, this)
        } else {
            listener!!.onAddClicked(this)
        }
    }

    fun setContact(contact: Contact?) {
        this.contact = contact
        update()
    }

    private fun update() {
        if (hasContact()) {
            apply()
            return
        }
        clear()
    }

    fun hasContact(): Boolean {
        return contact != null
    }

    @UiThread
    private fun clear() {
        post { setImageResource(R.drawable.ic_add) }
    }

    @UiThread
    private fun apply() {
        val drawable = if (contact!!.hasPhoto()) contact!!.roundedPhoto(context) else contact!!.letterDrawable()
        post { setImageDrawable(drawable) }
    }

    interface OnClickListener {
        fun onCallClicked(contact: Contact?, view: View?)
        fun onAddClicked(view: View?)
    }
}